File: single_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def is_empty(self):
        return self.head is None

    def length(self):
        current = self.head
        count = 0
        while current is not None:
            count += 1
            current = current.next
        return count

    def add_first(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def add_last(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        current = self.head
        while current.next is not None:
            current = current.next
        current.next = new_node

    def add_at_position(self, data, position):
        if position < 0 or position > self.length():
            raise ValueError("Invalid position")
        if position == 0:
            self.add_first(data)
            return
        new_node = Node(data)
        current = self.head
        for i in range(position - 1):
            current = current.next
        new_node.next = current.next
        current.next = new_node

    def remove(self, data):
        if self.head is None:
            return
        if self.head.data == data:
            self.head = self.head.next
            return
        current = self.head
        while current.next is not None:
            if current.next.data == data:
                current.next = current.next.next
                return
            current = current.next

    def remove_at_position(self, position):
        if position < 0 or position >= self.length():
            raise ValueError("Invalid position")
        if position == 0:
            self.head = self.head.next
            return
        current = self.head
        for i in range(position - 1):
            current = current.next
        current.next = current.next.next

    def search(self, data):
        current = self.head
        while current is not None:
            if current.data == data:
                return True
            current = current.next
        return False

    def get_position(self, data):
        current = self.head
        position = 0
        while current is not None:
            if current.data == data:
                return position
            position += 1
            current = current.next
        return -1

    def display(self):
        current = self.head
        while current is not None:
            print(current.data, end=' ')
            current = current.next
        print()

File: single_linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    int data;
    struct Node *next;
} node;

node *head = NULL;

void print_list(node *head)
{
    node *tmp = head;
    while(tmp != NULL)
    {
        printf("%d\n", tmp->data);
        tmp = tmp->next;
    }
}

void add_node(node **head, int n)
{
    node *new_node = malloc(sizeof(node));
    new_node->data = n;
    new_node->next = *head;
    *head = new_node;
}
node* search_list(node* head, int n)
{
    node* tmp = head;
    while (tmp != NULL)
    {
        if (tmp->data == n)
            return tmp;
        tmp = tmp->next;
    }
    return NULL;
}

void add_node_at_end(node** head, int n)
{
    node* new_node = malloc(sizeof(node));
    new_node->data = n;
    new_node->next = NULL;

    if (*head == NULL)
    {
        *head = new_node;
        return;
    }

    node* tmp = *head;
    while (tmp->next != NULL)
        tmp = tmp->next;
    tmp->next = new_node;
}

void add_node_in_middle(node* head, int pos, int n)
{
    node* new_node = malloc(sizeof(node));
    new_node->data = n;

    node* tmp = head;
    int i = 1;
    while (tmp != NULL && i < pos - 1)
    {
        tmp = tmp->next;
        i++;
    }

    if (tmp == NULL)
    {
        printf("Position not found in the list\n");
        return;
    }

    new_node->next = tmp->next;
    tmp->next = new_node;
}

void reverse_list(node** head)
{
    node* prev = NULL;
    node* current = *head;
    node* next = NULL;

    while (current != NULL)
    {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }

    *head = prev;
}
int main()
{
    int d;
    while (1)
    {
        printf("Please enter any number (or -1 to stop): ");
        scanf("%d", &d);
        
        if (d == -1)
            break;

        add_node(&head, d);
    }

    printf("Original List:\n");
    print_list(head);

    printf("\nSearching the list for value 5:\n");
    node* searched_node = search_list(head, 5);
    if (searched_node != NULL)
        printf("Node found with value %d\n", searched_node->data);
    else
        printf("Node not found\n");

    printf("\nAdding a new node with value 10 at the end:\n");
    add_node_at_end(&head, 10);
    print_list(head);

    printf("\nAdding a new node with value 20 at position 3:\n");
    add_node_in_middle(head, 3, 20);
    print_list(head);

    printf("\nReversing the list:\n");
    reverse_list(&head);
    print_list(head);

    return 0;
}
